compiler/noirc_evaluator/src/ssa/parser/into_ssa.rs

```diff
@@ -0,0 +1,309 @@
+use std::collections::HashMap;
+
+use im::Vector;
+
+use crate::ssa::{
+    function_builder::FunctionBuilder,
+    ir::{basic_block::BasicBlockId, function::FunctionId, value::ValueId},
+};
+
+use super::{
+    Identifier, ParsedBlock, ParsedFunction, ParsedInstruction, ParsedSsa, ParsedTerminator,
+    ParsedValue, RuntimeType, Ssa, SsaError,
+};
+
+impl ParsedSsa {
+    pub(crate) fn into_ssa(mut self) -> Result<Ssa, SsaError> {
+        let mut translator = Translator::new(&mut self)?;
+
+        for function in self.functions {
+            translator.translate_function(function)?;
+        }
+
+        Ok(translator.finish())
+    }
+}
+
+struct Translator {
+    builder: FunctionBuilder,
+
+    /// Maps function names to their IDs
+    functions: HashMap<String, FunctionId>,
+
+    /// Maps block names to their IDs
+    blocks: HashMap<FunctionId, HashMap<String, BasicBlockId>>,
+
+    /// Maps variable names to their IDs
+    variables: HashMap<FunctionId, HashMap<String, ValueId>>,
+}
+
+impl Translator {
+    fn new(parsed_ssa: &mut ParsedSsa) -> Result<Self, SsaError> {
+        let main_function = parsed_ssa.functions.remove(0);
+        let main_id = FunctionId::new(0);
+        let mut builder = FunctionBuilder::new(main_function.external_name.clone(), main_id);
+        builder.set_runtime(main_function.runtime_type);
+
+        // Map function names to their IDs so calls can be resolved
+        let mut function_id_counter = 1;
+        let mut functions = HashMap::new();
+        for function in &parsed_ssa.functions {
+            let function_id = FunctionId::new(function_id_counter);
+            function_id_counter += 1;
+
+            functions.insert(function.internal_name.clone(), function_id);
+        }
+
+        let mut translator =
+            Self { builder, functions, variables: HashMap::new(), blocks: HashMap::new() };
+        translator.translate_function_body(main_function)?;
```

Comment:
Why is it that main is translated here, but all the other functions are done in `into_ssa` above? I was wondering why `main` is not processed twice when I saw the `.remove(0)`. Would it be equally correct to loop through all the rest here?

Reply:
Good question! I tried to make it like you said, but it ends to a panic.

The reason is that it seems `FunctionBuilder` needs to be created with the main function. Once you do that you can call `set_runtime` to set the main function's runtime.

Then other functions need to be created with `new_function` or `new_brillig_function`... so dealing with the main function and the rest of the functions has to be done in different ways.

Maybe to have things a bit more uniform, we could change `FunctionBuilder` to be created without an initial function. Then you can create new functions and trying to insert an instruction before creating an initial function would panic. I guess it's not done like that to avoid an `Option` that could panic. But... we could maybe do this in a separate PR! In any case this special handling is not that much code, so maybe it's fine.

Reply:
Right, yes, I saw that the builder needs a main, thanks for pointing out the `new_function` vs `new_brillig_function` as well; I didn't think that the special handling of the first function was to be avoided, just found it odd that the main is _removed_ from the mutable `parsed_ssa`, and then the translator is returned to be fed more: at that point it looks like the only reasonable thing to do with the `parsed_ssa` is to consume all of its functions anyway, so I was wondering why it's not done in `new`. But it's all private, so ultimately doesn't matter that much, it was just like "why is it a mutable ref?" when I first saw it.

Reply:
>  I guess it's not done like that to avoid an Option that could panic. But... we could maybe do this in a separate PR! 

It's done for that reason and to avoid the Option entirely so that it is somewhat faster not having to unwrap/expect the current function on every value/instruction insertion.

Reply:
I would perhaps do it like 
```rust
impl Translator {
  pub fn translate(mut ssa: ParsedSsa) -> Result<Ssa, SsaError> {
    let mut t = Self::new(&mut ssa);
    for f in ssa.function { 
      t.translate_function(f);
    }
    t.finish()
  }
}
```
And then there is no question about who owns the `ssa` during construction.

Reply:
Ah, good point. I also thought it wasn't correct to have `new` modify `ParsedSsa`, then the caller doing more things. Making it a method of `Translator` makes more sense 👍

Reply:
(but I wasn't sure how to exactly improve it)

Reply:
Done! I also added a few comments and renamed a function.